crypto/evp/pkey_kdf.c

```diff
@@ -12,130 +12,187 @@
     case EVP_PKEY_CTRL_TLS_SEED:
-        cmd = EVP_KDF_CTRL_ADD_TLS_SEED;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_SEED;
         break;
     case EVP_PKEY_CTRL_HKDF_KEY:
-        cmd = EVP_KDF_CTRL_SET_KEY;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_KEY;
         break;
     case EVP_PKEY_CTRL_HKDF_INFO:
-        cmd = EVP_KDF_CTRL_ADD_HKDF_INFO;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_INFO;
         break;
     case EVP_PKEY_CTRL_HKDF_MODE:
-        cmd = EVP_KDF_CTRL_SET_HKDF_MODE;
+        cmd = T_INT;
+        name = OSSL_KDF_PARAM_MODE_INT;
         break;
     case EVP_PKEY_CTRL_SCRYPT_N:
-        cmd = EVP_KDF_CTRL_SET_SCRYPT_N;
+        cmd = T_UINT64;
+        name = OSSL_KDF_PARAM_SCRYPT_N;
         break;
     case EVP_PKEY_CTRL_SCRYPT_R:
-        cmd = EVP_KDF_CTRL_SET_SCRYPT_R;
+        cmd = T_UINT64; /* Range checking occurs on the provider side */
+        name = OSSL_KDF_PARAM_SCRYPT_R;
         break;
     case EVP_PKEY_CTRL_SCRYPT_P:
-        cmd = EVP_KDF_CTRL_SET_SCRYPT_P;
+        cmd = T_UINT64; /* Range checking occurs on the provider side */
+        name = OSSL_KDF_PARAM_SCRYPT_P;
         break;
     case EVP_PKEY_CTRL_SCRYPT_MAXMEM_BYTES:
-        cmd = EVP_KDF_CTRL_SET_MAXMEM_BYTES;
+        cmd = T_UINT64;
+        name = OSSL_KDF_PARAM_SCRYPT_MAXMEM;
         break;
     default:
         return -2;
     }
 
     switch (cmd) {
-    case EVP_KDF_CTRL_SET_PASS:
-    case EVP_KDF_CTRL_SET_SALT:
-    case EVP_KDF_CTRL_SET_KEY:
-    case EVP_KDF_CTRL_SET_TLS_SECRET:
-    case EVP_KDF_CTRL_ADD_TLS_SEED:
-    case EVP_KDF_CTRL_ADD_HKDF_INFO:
-        return EVP_KDF_ctrl(kctx, cmd, (const unsigned char *)p2, (size_t)p1);
-
-    case EVP_KDF_CTRL_SET_MD:
-        return EVP_KDF_ctrl(kctx, cmd, (const EVP_MD *)p2);
-
-    case EVP_KDF_CTRL_SET_HKDF_MODE:
-        return EVP_KDF_ctrl(kctx, cmd, (int)p1);
-
-    case EVP_KDF_CTRL_SET_SCRYPT_R:
-    case EVP_KDF_CTRL_SET_SCRYPT_P:
-        u64_value = *(uint64_t *)p2;
-        if (u64_value > UINT32_MAX) {
-            EVPerr(EVP_F_PKEY_KDF_CTRL, EVP_R_PARAMETER_TOO_LARGE);
-            return 0;
-        }
+    case T_OCTET_STRING:
+        *p = OSSL_PARAM_construct_octet_string(name, (unsigned char *)p2,
+                                               (size_t)p1);
+        break;
 
-        return EVP_KDF_ctrl(kctx, cmd, (uint32_t)u64_value);
+    case T_DIGEST:
+        mdname = EVP_MD_name((const EVP_MD *)p2);
+        *p = OSSL_PARAM_construct_utf8_string(name, (char *)mdname,
+                                              strlen(mdname) + 1);
+        break;
 
-    case EVP_KDF_CTRL_SET_SCRYPT_N:
-    case EVP_KDF_CTRL_SET_MAXMEM_BYTES:
-        return EVP_KDF_ctrl(kctx, cmd, *(uint64_t *)p2);
+        /*
+         * These are special because the helper macros pass a pointer to the
+         * stack, so a local copy is required.
+         */
+    case T_INT:
+        pkctx->ints[pkctx->pidx] = *(int *)p2;
```

Comment:
.... but then over here, you're using `p2` ?

Reply:
I think this is an error when I introduced the int saving.

I'm surprised that the tests passed on my machine (they aren't passing on the build ones).